File: frontend/lib/security/data-encryption.ts
import {
  createCipheriv,
  createDecipheriv,
  createHmac,
  randomBytes,
} from "node:crypto";

const ENVELOPE_PREFIX = "enc:v1";

type EncryptionConfig = {
  currentKey: Buffer | null;
  previousKey: Buffer | null;
  keyId: string;
};

let cachedConfig: EncryptionConfig | null = null;

function parseKey(raw: string): Buffer {
  const value = raw.trim();
  if (!value) {
    throw new Error("Encryption key cannot be empty.");
  }

  if (/^[0-9a-fA-F]+$/.test(value) && value.length % 2 === 0) {
    const decoded = Buffer.from(value, "hex");
    if (decoded.length === 32) return decoded;
  }

  const padded = value + "=".repeat((4 - (value.length % 4)) % 4);
  const decoded = Buffer.from(padded, "base64");
  if (decoded.length === 32) return decoded;

  throw new Error("Data encryption key must decode to exactly 32 bytes.");
}

function getConfig(): EncryptionConfig {
  if (cachedConfig) {
    return cachedConfig;
  }

  const currentRaw = process.env.DATA_ENCRYPTION_KEY_CURRENT?.trim() ?? "";
  const previousRaw = process.env.DATA_ENCRYPTION_KEY_PREVIOUS?.trim() ?? "";
  const keyId = process.env.DATA_ENCRYPTION_KEY_ID?.trim() || "k1";

  cachedConfig = {
    currentKey: currentRaw ? parseKey(currentRaw) : null,
    previousKey: previousRaw ? parseKey(previousRaw) : null,
    keyId,
  };
  return cachedConfig;
}

function base64urlEncode(raw: Buffer): string {
  return raw
    .toString("base64")
    .replace(/\+/g, "-")
    .replace(/\//g, "_")
    .replace(/=+$/g, "");
}

function base64urlDecode(raw: string): Buffer {
  const normalized = raw.replace(/-/g, "+").replace(/_/g, "/");
  const padded = normalized + "=".repeat((4 - (normalized.length % 4)) % 4);
  return Buffer.from(padded, "base64");
}

function blindIndexKey(key: Buffer): Buffer {
  return createHmac("sha256", key).update("blind-index:v1").digest();
}

export function isDataEncryptionEnabled(): boolean {
  return Boolean(getConfig().currentKey);
}

export function encryptValue(plaintext: string | null): string | null {
  if (plaintext === null) {
    return null;
  }

  const config = getConfig();
  if (!config.currentKey) {
    return null;
  }

  const nonce = randomBytes(12);
  const cipher = createCipheriv("aes-256-gcm", config.currentKey, nonce);
  const ciphertext = Buffer.concat([cipher.update(plaintext, "utf8"), cipher.final()]);
  const tag = cipher.getAuthTag();
  const payload = base64urlEncode(Buffer.concat([nonce, ciphertext, tag]));
  return `${ENVELOPE_PREFIX}:${config.keyId}:${payload}`;
}

export function decryptValue(ciphertext: string | null): string | null {
  if (ciphertext === null) {
    return null;
  }

  if (!ciphertext.startsWith(`${ENVELOPE_PREFIX}:`)) {
    return ciphertext;
  }

  const parts = ciphertext.split(":", 4);
  if (parts.length !== 4) {
    throw new Error("Invalid encrypted value format.");
  }

  const [, , embeddedKeyId, payload] = parts;
  const blob = base64urlDecode(payload);
  if (blob.length < 12 + 16) {
    throw new Error("Encrypted payload is too short.");
  }

  const nonce = blob.subarray(0, 12);
  const tag = blob.subarray(blob.length - 16);
  const encrypted = blob.subarray(12, blob.length - 16);

  const config = getConfig();
  if (!config.currentKey) {
    throw new Error("Encrypted data found but DATA_ENCRYPTION_KEY_CURRENT is not configured.");
  }

  const keyCandidates: Buffer[] = [];
  if (embeddedKeyId === config.keyId) {
    keyCandidates.push(config.currentKey);
    if (config.previousKey) keyCandidates.push(config.previousKey);
  } else {
    if (config.previousKey) keyCandidates.push(config.previousKey);
    keyCandidates.push(config.currentKey);
  }

  let lastError: unknown = null;
  for (const key of keyCandidates) {
    try {
      const decipher = createDecipheriv("aes-256-gcm", key, nonce);
      decipher.setAuthTag(tag);
      const decrypted = Buffer.concat([decipher.update(encrypted), decipher.final()]);
      return decrypted.toString("utf8");
    } catch (error) {
      lastError = error;
    }
  }

  const error = new Error("Failed to decrypt encrypted value with configured keys.");
  (error as Error & { cause?: unknown }).cause = lastError;
  throw error;
}

export function decryptWithFallback(
  ciphertext: string | null,
  plaintextFallback: string | null
): string | null {
  if (ciphertext) {
    try {
      return decryptValue(ciphertext);
    } catch {
      return plaintextFallback;
    }
  }
  return plaintextFallback;
}

export function blindIndex(value: string | null): string | null {
  if (value === null) return null;

  const config = getConfig();
  if (!config.currentKey) {
    return null;
  }

  return createHmac("sha256", blindIndexKey(config.currentKey)).update(value, "utf8").digest("hex");
}

export function blindIndexCandidates(value: string | null): string[] {
  if (value === null) return [];

  const config = getConfig();
  if (!config.currentKey) {
    return [];
  }

  const current = createHmac("sha256", blindIndexKey(config.currentKey))
    .update(value, "utf8")
    .digest("hex");
  if (!config.previousKey) {
    return [current];
  }

  const previous = createHmac("sha256", blindIndexKey(config.previousKey))
    .update(value, "utf8")
    .digest("hex");
  return previous === current ? [current] : [current, previous];
}

File: backend/app/security/data_encryption.py
"""
Application-layer field encryption helpers.

Envelope format:
    enc:v1:<keyId>:<base64url(nonce + ciphertext)>
"""
from __future__ import annotations

import base64
import hashlib
import hmac
import os
from dataclasses import dataclass
from functools import lru_cache
from typing import Optional

from cryptography.hazmat.primitives.ciphers.aead import AESGCM


_ENVELOPE_PREFIX = "enc:v1"


@dataclass(frozen=True)
class _EncryptionConfig:
    current_key: Optional[bytes]
    previous_key: Optional[bytes]
    key_id: str

    @property
    def enabled(self) -> bool:
        return self.current_key is not None


def _parse_key(raw: str) -> bytes:
    candidate = raw.strip()
    if not candidate:
        raise ValueError("Encryption key cannot be empty.")

    # Support hex keys for operational convenience.
    if all(ch in "0123456789abcdefABCDEF" for ch in candidate) and len(candidate) % 2 == 0:
        decoded = bytes.fromhex(candidate)
        if len(decoded) == 32:
            return decoded

    padded = candidate + ("=" * ((4 - len(candidate) % 4) % 4))
    try:
        decoded = base64.urlsafe_b64decode(padded.encode("utf-8"))
        if len(decoded) == 32:
            return decoded
    except Exception as exc:  # pragma: no cover - explicit error below
        raise ValueError("Invalid base64 data encryption key.") from exc

    raise ValueError("Data encryption key must decode to exactly 32 bytes.")


@lru_cache(maxsize=1)
def _load_config() -> _EncryptionConfig:
    current_raw = os.getenv("DATA_ENCRYPTION_KEY_CURRENT", "").strip()
    previous_raw = os.getenv("DATA_ENCRYPTION_KEY_PREVIOUS", "").strip()
    key_id = os.getenv("DATA_ENCRYPTION_KEY_ID", "k1").strip() or "k1"

    current_key = _parse_key(current_raw) if current_raw else None
    previous_key = _parse_key(previous_raw) if previous_raw else None

    return _EncryptionConfig(
        current_key=current_key,
        previous_key=previous_key,
        key_id=key_id,
    )


def is_data_encryption_enabled() -> bool:
    return _load_config().enabled


def _urlsafe_b64encode(raw: bytes) -> str:
    return base64.urlsafe_b64encode(raw).decode("utf-8").rstrip("=")


def _urlsafe_b64decode(raw: str) -> bytes:
    padded = raw + ("=" * ((4 - len(raw) % 4) % 4))
    return base64.urlsafe_b64decode(padded.encode("utf-8"))


def encrypt_value(plaintext: Optional[str]) -> Optional[str]:
    if plaintext is None:
        return None

    config = _load_config()
    if not config.enabled:
        return None

    nonce = os.urandom(12)
    ciphertext = AESGCM(config.current_key).encrypt(
        nonce=nonce,
        data=plaintext.encode("utf-8"),
        associated_data=None,
    )
    payload = _urlsafe_b64encode(nonce + ciphertext)
    return f"{_ENVELOPE_PREFIX}:{config.key_id}:{payload}"


def decrypt_value(ciphertext: Optional[str]) -> Optional[str]:
    if ciphertext is None:
        return None

    if not ciphertext.startswith(f"{_ENVELOPE_PREFIX}:"):
        # Backward compatibility with plaintext rows.
        return ciphertext

    parts = ciphertext.split(":", 3)
    if len(parts) != 4:
        raise ValueError("Invalid encrypted value format.")

    _enc, _version, embedded_key_id, payload = parts
    blob = _urlsafe_b64decode(payload)
    if len(blob) < 13:
        raise ValueError("Encrypted payload is too short.")

    nonce = blob[:12]
    encrypted = blob[12:]

    config = _load_config()
    if not config.enabled:
        raise ValueError("Encrypted data found but DATA_ENCRYPTION_KEY_CURRENT is not configured.")

    candidate_keys = []
    if embedded_key_id == config.key_id:
        candidate_keys.append(config.current_key)
        if config.previous_key:
            candidate_keys.append(config.previous_key)
    else:
        if config.previous_key:
            candidate_keys.append(config.previous_key)
        candidate_keys.append(config.current_key)

    last_error = None
    for key in candidate_keys:
        try:
            plaintext = AESGCM(key).decrypt(
                nonce=nonce,
                data=encrypted,
                associated_data=None,
            )
            return plaintext.decode("utf-8")
        except Exception as exc:  # pragma: no cover - final error below
            last_error = exc

    raise ValueError("Failed to decrypt encrypted value with configured keys.") from last_error


def decrypt_with_fallback(ciphertext: Optional[str], plaintext_fallback: Optional[str]) -> Optional[str]:
    if ciphertext:
        return decrypt_value(ciphertext)
    return plaintext_fallback


def _blind_index_key(raw_key: bytes) -> bytes:
    return hmac.new(raw_key, b"blind-index:v1", hashlib.sha256).digest()


def blind_index(value: Optional[str]) -> Optional[str]:
    if value is None:
        return None

    config = _load_config()
    if not config.enabled:
        return None

    key = _blind_index_key(config.current_key)
    return hmac.new(key, value.encode("utf-8"), hashlib.sha256).hexdigest()


def blind_index_candidates(value: Optional[str]) -> list[str]:
    if value is None:
        return []

    config = _load_config()
    if not config.enabled:
        return []

    candidates: list[str] = []
    current = hmac.new(
        _blind_index_key(config.current_key),
        value.encode("utf-8"),
        hashlib.sha256,
    ).hexdigest()
    candidates.append(current)

    if config.previous_key:
        previous = hmac.new(
            _blind_index_key(config.previous_key),
            value.encode("utf-8"),
            hashlib.sha256,
        ).hexdigest()
        if previous != current:
            candidates.append(previous)

    return candidates


def reset_encryption_config_cache() -> None:
    _load_config.cache_clear()
